moby/moby
# Add IP_NF_MANGLE to check-config.sh

Add IP_NF_MANGLE to "Generally Required" kernel features, since it appears to be necessary for Docker Swarm to work.

Closes #46636

**- What I did**
Used the `check-config.sh` script for my custom Linux to be compatible with Docker.  And it mostly works, except connecting to local services in swarm mode! 

I tested on a single-node swarm cluster and started a simple web service:
```
docker swarm init --default-addr-pool 10.22.0.0/16
docker service create --publish published=8088,target=80 --name=www nginx:1.25.2-alpine
curl 127.0.0.1:8088
``` 
The last command will fail to connect to the published port: `curl: (7) Failed to connect to 127.0.0.1 port 8088 after 0 ms: Error`


**- How I did it**
I isolated the cause for the connection refused to a single kernel module, using lots of `lsmod` and `diff` between a working system and my broken system, and moving various kernel modules out of the way and back again. It took some time.

**- How to verify it**
First establish a failing test case:
1. Boot into a Linux that starts a Docker Engine but any swarm nodes.
2. `sudo mv /lib/modules/6.1.51/kernel/net/ipv4/netfilter/iptable_mangle.ko{,.disabled}` (or equivalent) to move kernel modules out of the way.
3. `docker swarm init --default-addr-pool 10.22.0.0/16`
4. `docker service create --publish published=8088,target=80 --name=www nginx:1.25.2-alpine`
6. `curl 127.0.0.1:8088`
Expected: Connected, HTML is shown .
Actual: Connection failed.
Run `lsmod | grep iptable_mangle` and confirm the module isn't loaded.

Cleanup:
1. `docker service rm www`
2. `docker swarm leave --force`


Now confirm that the module is indeed required by swarm:
1. `sudo mv /lib/modules/6.1.51/kernel/net/ipv4/netfilter/iptable_mangle.ko{.disabled,}` to reinstall kernel module
2. `docker swarm init --default-addr-pool 10.22.0.0/16`
3. `lsmod | grep iptable_mangle` and confirm the module is loaded but unused.
4. `docker service create --publish published=8088,target=80 --name=www nginx:1.25.2-alpine`
5. Run `lsmod | grep iptable_mangle` and confirm the module is in fact loaded and in use.
6. `curl 127.0.0.1:8088`
Expected/Actual: `iptable_mangle` is listed, HTML is shown.



**- Description for the changelog**
Add IP_NF_MANGLE to the list of "Generally Required" because it is required by Swarm.


**- A picture of a cute animal (not mandatory but encouraged)**

```
                               __
                              /\_\
                             _\/_/
                            /\_\
                            \/_/_
                              /\_\
                              \/_/
                             /\_\               .
                             \/_/              ::
                              .:::::::::::::..::::.
                              :::::::::::::::::
                              ;;;;;;.:::::::::
                               ;;;;;;:::::::
                                 ;;;:::::

                        https://lightwhale.asklandd.dk

```

Signed-off-by: Stephan Henningsen <stephan+github@asklandd.dk>
